### ISSUE TITLE
html5.js should use non-nullable types

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#860]
Several types specified in html5.js should be indicated as non-nullable types.  For example, here is an error I get (with "new type inference" compiler option) which should not occur:

```
src/lab/graph/StandardAxes.js:296: WARNING - Attempt to use nullable type TextMetrics|null.
      var textWidth = context.measureText(s).width;
                      ^
```

where `context` is defined by:

```
@param {!CanvasRenderingContext2D} context 
```

The definition referenced in html5.js is http://www.w3.org/TR/2dcontext/#textmetrics.  That page says that "Interfaces are defined in terms of [Web IDL](http://heycam.github.io/webidl/)".  Looking at that page, in the [section about Nullable Types](http://heycam.github.io/webidl/#idl-nullable-type) we read:

> A nullable type is an IDL type constructed from an existing type (called the inner type), which just allows the additional value null to be a member of its set of values. Nullable types are represented in IDL by placing a U+003F QUESTION MARK ("?") character after an existing type. The inner type must not be any, another nullable type, or a union type that itself has includes a nullable type or has a dictionary type as one of its flattened member types.

The [w3.org definition of CanvasRenderingContext2D](http://www.w3.org/TR/2dcontext/#canvasrenderingcontext2d) shows that measureText returns a non-nullable type:

```
interface CanvasRenderingContext2D {
 // code omitted...
 TextMetrics measureText(DOMString text);
// code omitted...
};
interface TextMetrics {
  readonly attribute double width;
};
```

The same argument applies to other return types in CanvasRenderingContext2D such as CanvasGradient, CanvasPattern, and ImageData.

**Comment by myphysicslab** [comment]
Another property to perhaps make non-nullable is `window.localStorage`

```
src/lab/util/Terminal.js:412: WARNING - Attempt to use nullable type Storage|null.
  window.localStorage.removeItem(this.pageKey());
  ^
```
This would be in the extern webstorage.js:
```javascript
/**
 * Window implements WindowLocalStorage
 *
 * @type {Storage}
 */
Window.prototype.localStorage;
```

**Comment by myphysicslab** [comment]
Another property to consider for making non-nullable: `navigator`
```
src/lab/util/UtilityCore.js:228: WARNING - Attempt to use nullable type Navigator|null.
  return navigator.userAgent.match(/.*Chrome.*/) != null;
         ^
```
Or can `navigator` be null in older browsers?

**Comment by dimvar** [comment]
This is really an externs issue, not an NTI issue, so I'm removing the label.

**Comment by Dominator008** [comment]
@myphysicslab Care to send a PR?

**Comment by chrismiddleton** [comment]
I went through the html5.js file and found all the ones that seem to be missing a "!" prepended. This list might need double-checking, but I think they're all valid misses. The one that brought me here originally was drawImage missing a "!" in its first argument

    @param {!Node} n
    Node.prototype.contains

    @return {!CanvasGradient}
    CanvasRenderingContext2D.prototype.createRadialGradient

	@param {!HTMLImageElement|!HTMLCanvasElement} image
	@return {!CanvasPattern}
	CanvasRenderingContext2D.prototype.createPattern
	
	@return {!TextMetrics}
	CanvasRenderingContext2D.prototype.measureText
	
	@param {!HTMLImageElement|!HTMLCanvasElement|!Image|!HTMLVideoElement} image
	CanvasRenderingContext2D.prototype.drawImage
	
	@param {!ImageData} imagedata
	CanvasRenderingContext2D.prototype.putImageData
	
	@return {!Array<number>}
	CanvasRenderingContext2D.prototype.getLineDash
	
 	@param {!Array<number>} segments
	CanvasRenderingContext2D.prototype.setLineDash;
	
	@param {!Uint8ClampedArray|number}
	function ImageData(dataOrWidth, widthOrHeight, opt_height) {}
	
	@type {!Uint8ClampedArray}
	ImageData.prototype.data;
	
	// these first two actually are supposed to be optional according to the spec but weren't
	// the third needs to allow undefined too
	@param {((function(!SQLTransaction) : void)?)=} callback
	@param {((function(!SQLError) : void)?)=} errorCallback
	@param {(Function?)=} successCallback
	Database.prototype.changeVersion
	
	@type {!SQLResultSetRowList}
	SQLResultSet.prototype.rows;
	
	
	@type {!WorkerGlobalScope}
	WorkerGlobalScope.prototype.self;

	@type {!WorkerLocation}
	WorkerGlobalScope.prototype.location;
	
	@type {!TimeRanges}
	HTMLMediaElement.prototype.buffered;
	HTMLMediaElement.prototype.played;
	HTMLMediaElement.prototype.seekable;
	
	@return {!TextTrack}
	HTMLMediaElement.prototype.addTextTrack
	
	/** @type {!TextTrackList} */
	HTMLMediaElement.prototype.textTracks;
	
	 * @param {!TextTrackCue} cue
	TextTrack.prototype.addCue = function(cue) {};

	 * @param {!TextTrackCue} cue
	TextTrack.prototype.removeCue = function(cue) {};
	
	/** @type {Array<string>!} */
	DataTransfer.prototype.types;

	
	/** @type {!FileList} */
	DataTransfer.prototype.files;
	
	@param {!HTMLElement} elem
	DataTransfer.prototype.addElement
	
	@param {!HTMLElement} img
	DataTransfer.prototype.setDragImage
	
	@param {string|!ArrayBuffer|!ArrayBufferView} data
	WebSocket.prototype.send
	
	@param {!XMLHttpRequestUpload}
	XMLHttpRequest.prototype.upload;
	
	@type {!Object<string,string>}
	HTMLElement.prototype.dataset;
	
	@type {!NodeList<!HTMLLabelElement>}
	HTMLButtonElement.prototype.labels;
	
	@type {!NodeList<!HTMLLabelElement>}
	HTMLInputElement.prototype.labels;
	
	@type {!NodeList<!HTMLLabelElement>}
	HTMLSelectElement.prototype.labels;
	
	@type {!NodeList<!HTMLLabelElement>}
	HTMLTextAreaElement.prototype.labels;
	
	/** @type {!Node} */
	MutationRecord.prototype.target;

	/** @type {!NodeList<!Node>} */
	MutationRecord.prototype.addedNodes;

	/** @type {!NodeList<!Node>} */
	MutationRecord.prototype.removedNodes;
	
	@param {!Node} target
	MutationObserver.prototype.observe
	
	@const {!NodeList<!HTMLLabelElement>}
	HTMLOutputElement.prototype.labels;
	
	@type {!NodeList<!Node>}
	HTMLProgressElement.prototype.labels;
	
	/** @const {!TextTrack} */
	HTMLTrackElement.prototype.track;
	
	/** @type {!NodeList<!Node>} */
	HTMLMeterElement.prototype.labels;
	
	 * @type {!MimeTypeArray}
	Navigator.prototype.mimeTypes;
	
	 * @type {!PluginArray}
	Navigator.prototype.plugins;

**Comment by brad4d** [comment]
FYI, fixing return types and property values to be non-nullable will probably be an easy change to land, since existing code that expects nullable values won't break if they never get a null.
Fixing parameter types to be non-nullable probably will break some existing code, though, since it'll need to be modified to make sure it isn't ever passing a null or looks to the compiler like it could.

**Comment by myphysicslab** [comment]
Please see issue #2425 where I've started a pull request for some of these.  I can add some or all of the changes that @chrismiddleton listed to that pull request if desired.

I appreciate @brad4d comment that changing parameter types can break existing code, so it needs discussion.  I think that since HTML5 is a newer interface there will be fewer users affected.  Plus this is exactly the type of checking that closure compiler is useful for.  So I would vote for getting these changes in now, because in future it will become impossible.  How is a decision like this made on this project?

The specs are not always obvious about when a function can return null.  For example see the spec for createPattern which casually mentions that null can be returned in the description: "If it returns bad, then return null and abort these steps". See
https://html.spec.whatwg.org/multipage/scripting.html#dom-context-2d-createpattern

**Comment by ChadKillingsworth** [comment]
> How is a decision like this made on this project?

By willingness of a Google Team member to update potentially hundreds of internal Google projects. The compiler project is not allowed to commit code which breaks ANY existing project at Google.

**Comment by myphysicslab** [comment]
Fixed by #2435.
